x-pack/plugins/endpoint/common/generate_data.ts

```diff
@@ -0,0 +1,410 @@
+    variant: 'Windows Server Release 2',
+  },
+];
+
+const Linux: OSFields[] = [];
+
+const Mac: OSFields[] = [];
+
+const OS: OSFields[] = [...Windows, ...Mac, ...Linux];
+
+const POLICIES: Array<{ name: string; id: string }> = [
+  // mapping name and ID as a sperate attribute makes query more complicated
+  // perhaps, combine them into a single field such that
+  // policy.namd_id = 'Default:C2A9093E-E289-4C0A-AA44-8C32A414FA7A'
+  {
+    name: 'Default',
+    id: '00000000-0000-0000-0000-000000000000',
+  },
+  {
+    name: 'With Eventing',
+    id: 'C2A9093E-E289-4C0A-AA44-8C32A414FA7A',
+  },
+];
+
+const FILE_OPERATIONS: string[] = ['creation', 'open', 'rename', 'execution', 'deletion'];
+
+// These are from the v1 schemas and aren't all valid ECS event categories, still in flux
+const OTHER_EVENT_CATEGORIES: string[] = ['driver', 'file', 'library', 'network', 'registry'];
+
+function randomN(n: number): number {
+  return Math.floor(Math.random() * n);
+}
+
+function* randomNGenerator(max: number, count: number) {
+  while (count > 0) {
+    yield randomN(max);
+    count--;
+  }
+}
+
+function randomArray<T>(lenghtLimit: number, generator: () => T): T[] {
+  const rand = randomN(lenghtLimit) + 1;
+  return [...Array(rand).keys()].map(generator);
+}
+
+function randomMac(): string {
+  return [...randomNGenerator(255, 6)].map(x => x.toString(16)).join('-');
+}
+
+function randomIP(): string {
+  return [10, ...randomNGenerator(255, 3)].map(x => x.toString()).join('.');
+}
+
+function randomVersion(): string {
+  return [6, ...randomNGenerator(10, 2)].map(x => x.toString()).join('.');
+}
+
+function randomChoice<T>(arg: T[]): T {
+  return arg[randomN(arg.length)];
+}
+
+function randomString(length: number): string {
+  return [...randomNGenerator(36, length)].map(x => x.toString(36)).join('');
+}
+
+function randomHostname(): string {
+  return `Host-${randomString(10)}`;
+}
+
+export function generateRelatedEvents(
+  node: EndpointEvent,
+  generator: EndpointDocGenerator,
+  numRelatedEvents = 10
+): EndpointEvent[] {
+  const ts = node['@timestamp'] + 1000;
+  const relatedEvents: EndpointEvent[] = [];
+  for (let i = 0; i < numRelatedEvents; i++) {
+    relatedEvents.push(
+      generator.generateEvent(
+        ts,
+        node.process.entity_id,
+        node.process.parent?.entity_id,
+        randomChoice(OTHER_EVENT_CATEGORIES)
+      )
+    );
+  }
+  return relatedEvents;
+}
+
+export function generateResolverTree(
+  root: EndpointEvent,
+  generator: EndpointDocGenerator,
+  generations = 2,
+  maxChildrenPerNode = 2,
+  relatedEventsPerNode = 3,
+  percentNodesWithRelated = 100,
+  percentChildrenTerminated = 100
+): EndpointEvent[] {
+  let events: EndpointEvent[] = [root];
+  let parents = [root];
+  let timestamp = root['@timestamp'];
+  for (let i = 0; i < generations; i++) {
+    const newParents: EndpointEvent[] = [];
+    parents.forEach(element => {
+      // const numChildren = randomN(maxChildrenPerNode);
```

Comment:
you might still be working on stuff, but probably can delete this

Reply:
oops yeah i need to replace the random generator with a seeded random generator so we can have determinism but also a simple way to get variety